Fix for #9524: fix recursive included decorator style propagation

level420 opened the pull request:
Replaced the faulty recursive decorator include algorithm.

See #9524.

level420 committed:
Fix for #9524: fix recursive included style propagation

Replaced the faulty recursive decorator include algorithm.

derrell commented on framework/source/class/qx/theme/manager/Decoration.js:
```diff
@@ -214,54 +214,28 @@ qx.Class.define("qx.theme.manager.Decoration",
+          recurseEntryInclude(theme.decorations[currentEntry.include]);
+        }
+        
+        // if the docorator is not set yet, clone the topmost 
```
if the *decorator* is not set...

derrell reviewed (changes requested):
Very minor. Just a typo. Sorry.

level420 committed:
Fix a typo

level420 committed:
Removed comment which applied to an older revision

Adolfrizer commented:
You changed the order of addition properties so an operator `in` will iterate them in this order:
`local style (input-focused) -> input-base -> input`
but the correct order is:
`input-base -> input -> local style (input-focused)`
I added local style after all others

level420 committed:
start recursion with the current decorator, copying styles over into an empty style object

level420 commented:
Please do not approve currently. Still in discussion. Thank you.

level420 commented:
@Adolfrizer Please have again a look into the PR. I've fixed one existing problem and simplified the algorithm even more.

The "trick" is, that recursion starts with the current decorator definition, in the example `input-focused` in function `recurseDecoratorInclude` where, if an `include` property exists, it recursively calls the function with the `input` decorator which recursively calls the function with the `input-base` decorator.

As the recursion ends here, the styles get copied first over from `input-base` then from `input` and last from `input-focused`. This should be the right order, beginning from the topmost include down to the current decorator.

Additionally I've removed the unneeded clone of the `decoration` entry which seems to me to be a relic from ancient code, as not decorator entry has an attribute called `decorator`.

As only the `style` object/hash is used when creating the decorator, there is no need for `entry.styles`.

level420 commented:
And also note that the PR will not include the debug message anymore. This is just to illustrate that copying the styles works in the right order. 

The `name` parameter of the function `recurseDecoratorInclude` could also be used to track if we have circular references in the decorators, thus things like `dec1.include = "dec2"` and `dec2.include = "dec1"`.

level420 committed:
Preserve precedence within the decorator hierarchy and sort properties before applying so that short cut properties like width come before explicit ones like widthLeft

level420 committed:
remove property sorting

level420 committed:
do not return an assignment

Adolfrizer reviewed (approved):
I think it's OK. I verified it locally

johnspackman commented:
full approval, merging thank you @level420 !

hkollmann commented:
@level420 @Adolfrizer : Can we close this issue?
